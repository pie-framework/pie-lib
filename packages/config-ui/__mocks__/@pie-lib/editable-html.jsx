import React from 'react';

export default () => <div>EditableHtml</div>;
