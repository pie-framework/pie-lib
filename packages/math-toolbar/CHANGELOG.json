[
  {
    "type": "fix",
    "scope": null,
    "subject": "rm child prepack in favour of root prepack",
    "merge": null,
    "header": "fix: rm child prepack in favour of root prepack",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "381d8d6485269b641d8d4655148966d2e70803be",
    "gitTags": "",
    "committerDate": "2019-03-15 09:51:11 +0000",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.9.13"
  },
  {
    "type": "chore",
    "scope": null,
    "subject": "get packages to call changelog function on prepack",
    "merge": null,
    "header": "chore: get packages to call changelog function on prepack",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "8828f01f2d16d29404ba837c1de40d09e7230ec5",
    "gitTags": "",
    "committerDate": "2019-03-14 12:40:05 +0000",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.9.10"
  },
  {
    "type": "chore",
    "scope": null,
    "subject": "call changelog on prepack",
    "merge": null,
    "header": "chore: call changelog on prepack",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "b672c012de096c55b765a765f758da3fc5df2f90",
    "gitTags": "",
    "committerDate": "2019-03-13 23:00:44 +0000",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.9.10"
  },
  {
    "type": "chore",
    "scope": null,
    "subject": "generate changelog.json instead of markdown",
    "merge": null,
    "header": "chore: generate changelog.json instead of markdown",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "eaa5019a11f70c53425d55856bd3b940237f8c7d",
    "gitTags": "",
    "committerDate": "2019-03-13 17:24:09 +0000",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.9.10"
  },
  {
    "type": "chore",
    "scope": null,
    "subject": "force version bump to avoid collision w/ next packages",
    "merge": null,
    "header": "chore: force version bump to avoid collision w/ next packages",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "a05f0fdb47f72d3a3d839cd70e96de6642a44c1e",
    "gitTags": "",
    "committerDate": "2019-03-12 21:55:40 +0000",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.9.10"
  },
  {
    "type": "feat",
    "scope": "math-toolbar",
    "subject": "configurable keypad for extra button variety",
    "merge": null,
    "header": "feat(math-toolbar): configurable keypad for extra button variety",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "f219ae19c18a4166ccc4f7d6e1d6b24b0b444fd7",
    "gitTags": "",
    "committerDate": "2019-02-20 12:01:12 +0200",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.9.0"
  },
  {
    "type": "fix",
    "scope": "math-toolbar",
    "subject": "styling fix for mathquill fractions",
    "merge": null,
    "header": "fix(math-toolbar): styling fix for mathquill fractions",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "35b0d2dc2ac4f0b0f3b6617603367f570e418f10",
    "gitTags": "",
    "committerDate": "2019-02-13 11:45:49 +0200",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.8.0"
  },
  {
    "type": "fix",
    "scope": null,
    "subject": "remove unnecessary mathquill wrappers in favor of @pie-lib/math-input and small styling fixes",
    "merge": null,
    "header": "fix: remove unnecessary mathquill wrappers in favor of @pie-lib/math-input and small styling fixes",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "b29bc9a50123944921ee65f1a561c32c080b5455",
    "gitTags": "",
    "committerDate": "2019-02-12 16:44:04 +0200",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.8.0"
  },
  {
    "type": "fix",
    "scope": null,
    "subject": "fraction button in HorizontalKeypad",
    "merge": null,
    "header": "fix: fraction button in HorizontalKeypad",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "0d5d89620b2f959468fc88c89e7f37d1d8e9040a",
    "gitTags": "",
    "committerDate": "2019-02-12 13:02:53 +0000",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.8.0"
  },
  {
    "type": "feat",
    "scope": null,
    "subject": "version bump",
    "merge": null,
    "header": "feat: version bump",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "fef98bda03c1e7169c8576028f7b6b8abccbd021",
    "gitTags": "",
    "committerDate": "2019-02-11 11:32:10 +0000",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.7.0"
  },
  {
    "type": "feat",
    "scope": "math-toolbar",
    "subject": "mathquill paranthesis styling fix for custom elements",
    "merge": null,
    "header": "feat(math-toolbar): mathquill paranthesis styling fix for custom elements",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "dd1b720aba42a91aa3360e79c6c425ce04e34175",
    "gitTags": "",
    "committerDate": "2019-02-07 12:50:10 +0200",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.8.0"
  },
  {
    "type": "feat",
    "scope": "math-toolbar",
    "subject": "line break support",
    "merge": null,
    "header": "feat(math-toolbar): line break support",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "6b55597100a404d9843a61b2edefbe681ff2f0d9",
    "gitTags": "",
    "committerDate": "2019-02-06 10:47:43 +0200",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.8.0"
  },
  {
    "type": "chore",
    "scope": "math-toolbar",
    "subject": "dependency version bump to pull in fixes",
    "merge": null,
    "header": "chore(math-toolbar): dependency version bump to pull in fixes",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "63627bfd5f87f00387f8ac30c211bc7e9deeba6b",
    "gitTags": "",
    "committerDate": "2019-02-05 18:43:17 +0200",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.8.0"
  },
  {
    "type": "feat",
    "scope": "math-inline",
    "subject": "small styling fixes and adjustments",
    "merge": null,
    "header": "feat(math-inline): small styling fixes and adjustments",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "3b2b1426231d233cf7f736a7fe26a96a16b0df08",
    "gitTags": "",
    "committerDate": "2019-02-05 18:30:12 +0200",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.8.0"
  },
  {
    "type": "fix",
    "scope": null,
    "subject": "version bump",
    "merge": null,
    "header": "fix: version bump",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "caaf960891fb17ad8e221849ce992381e4c2c31c",
    "gitTags": "",
    "committerDate": "2019-02-05 15:38:33 +0000",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.6.4"
  },
  {
    "type": "feat",
    "scope": null,
    "subject": "libs version bump",
    "merge": null,
    "header": "feat: libs version bump",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "8bedee111c84126beaefbfd5cf01bac5fb1b016e",
    "gitTags": "",
    "committerDate": "2019-01-30 22:14:28 +0000",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.6.0"
  },
  {
    "type": "fix",
    "scope": "general",
    "subject": "bumped versions of some packages",
    "merge": null,
    "header": "fix(general): bumped versions of some packages",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "a38c7b58270773c8a359f82a6c40f86c0f00456e",
    "gitTags": "",
    "committerDate": "2019-01-08 14:18:19 +0200",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.5.0"
  },
  {
    "type": "fix",
    "scope": "math-toolbar",
    "subject": "fixed test",
    "merge": null,
    "header": "fix(math-toolbar): fixed test",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "84adb1507b12ef87bb8eaaa59c0b87184d82dd17",
    "gitTags": "",
    "committerDate": "2019-01-08 11:44:20 +0200",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.5.0"
  },
  {
    "type": "chore",
    "scope": null,
    "subject": "fix conflict",
    "merge": null,
    "header": "chore: fix conflict",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "55181f1c118a0f640b7975f212b9adf454896110",
    "gitTags": "",
    "committerDate": "2018-12-18 13:07:02 +0000",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.4.0"
  },
  {
    "type": "fix",
    "scope": null,
    "subject": "version bump",
    "merge": null,
    "header": "fix: version bump",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "3dcbea969abef98945b73a73fb73174cecb99a11",
    "gitTags": "",
    "committerDate": "2018-12-18 12:59:14 +0000",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.3.2"
  },
  {
    "type": "feat",
    "scope": "math-toolbar",
    "subject": "version bumps",
    "merge": null,
    "header": "feat(math-toolbar): version bumps",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "e88dc9ae8d549a9c4a6f1183fe1d7d334d674fe6",
    "gitTags": "",
    "committerDate": "2018-12-18 14:39:32 +0200",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.4.0"
  },
  {
    "type": "feat",
    "scope": "math-toolbar",
    "subject": "answer block initial implementation",
    "merge": null,
    "header": "feat(math-toolbar): answer block initial implementation",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "2200ee4258e316e2f0d3903751e088cc30c50318",
    "gitTags": "",
    "committerDate": "2018-12-18 13:52:00 +0200",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.4.0"
  },
  {
    "type": "feat",
    "scope": "math-toolbar",
    "subject": "answer block column for keypad and related functionality",
    "merge": null,
    "header": "feat(math-toolbar): answer block column for keypad and related functionality",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "2fd700a128f6d513d8cfa51473c2f384b8948f9e",
    "gitTags": "",
    "committerDate": "2018-11-29 13:34:45 +0200",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.3.0"
  },
  {
    "type": "feat",
    "scope": "math-toolbar",
    "subject": "changed from ssh to https for the mathquill forked repo",
    "merge": null,
    "header": "feat(math-toolbar): changed from ssh to https for the mathquill forked repo",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "7364008088bb47419af4391c2bc5aa17173692b5",
    "gitTags": "",
    "committerDate": "2018-11-27 16:02:58 +0200",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.2.0"
  },
  {
    "type": "feat",
    "scope": "math-toolbar",
    "subject": "updated the mathquill dependency to use the forked repo",
    "merge": null,
    "header": "feat(math-toolbar): updated the mathquill dependency to use the forked repo",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "7cc7c0651f43a0867ff91321b32dcbe976b0fb7c",
    "gitTags": "",
    "committerDate": "2018-11-27 14:55:51 +0200",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.2.0"
  },
  {
    "type": "feat",
    "scope": "math-toolbar",
    "subject": "broke apart editable-html math toolbar into a separate package",
    "merge": null,
    "header": "feat(math-toolbar): broke apart editable-html math toolbar into a separate package",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "72c5a794fb650c8a033e706c6d588afb544522eb",
    "gitTags": "",
    "committerDate": "2018-11-22 19:03:24 +0200",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.1.0"
  },
  {
    "type": "fix",
    "scope": "math-toolbar",
    "subject": "z-index fix for math toolbar",
    "merge": null,
    "header": "fix(math-toolbar): z-index fix for math toolbar",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "ef50bb61cdf7c0670c2de789efa6062e3c988f4d",
    "gitTags": "",
    "committerDate": "2019-03-21 10:49:43 +0200",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.9.14"
  },
  {
    "type": "fix",
    "scope": "math-toolbar",
    "subject": "added logic to make sure answer blocks are not duplicates",
    "merge": null,
    "header": "fix(math-toolbar): added logic to make sure answer blocks are not duplicates",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "d8dd51338bb9901acaba996bd6e72815ef80078f",
    "gitTags": "",
    "committerDate": "2019-03-20 16:24:50 +0200",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.9.14"
  },
  {
    "type": "feat",
    "scope": null,
    "subject": "use @pie-framework/mathquill - our fork with extras",
    "merge": null,
    "header": "feat: use @pie-framework/mathquill - our fork with extras",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "7fc47a75b5153065e3f26ecce2a8e79daacc389e",
    "gitTags": "",
    "committerDate": "2019-04-03 22:05:15 +0100",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.10.0"
  },
  {
    "type": "feat",
    "scope": "math-toolbar",
    "subject": "no decimals support added",
    "merge": null,
    "header": "feat(math-toolbar): no decimals support added",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "39baa2ec9cbfe7adcacba68a43c581d8ee64cf1d",
    "gitTags": "",
    "committerDate": "2019-04-10 10:39:05 +0300",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@0.11.0"
  },
  {
    "type": "fix",
    "scope": "math-toolbar",
    "subject": "response block addition switched to %response% from embeded custom element",
    "merge": null,
    "header": "fix(math-toolbar): response block addition switched to %response% from embeded custom element",
    "body": null,
    "footer": "BREAKING CHANGE: response block addition now uses a different format",
    "notes": [
      {
        "title": "BREAKING CHANGE",
        "text": "response block addition now uses a different format"
      }
    ],
    "hash": "0eb6f94c8def870b495e0c96f531bdadf32a4c7b",
    "gitTags": " (origin/feat/math-inline-formula-v2)",
    "committerDate": "2019-04-22 14:18:02 +0300",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@1.0.0"
  },
  {
    "type": "feat",
    "scope": "math-input",
    "subject": "custom keys support and extended keypads",
    "merge": null,
    "header": "feat(math-input): custom keys support and extended keypads",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "34a428b6f6007d4757c7cfa66e187b24ddb085b9",
    "gitTags": "",
    "committerDate": "2019-04-25 16:16:24 +0300",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@1.1.0"
  },
  {
    "type": "feat",
    "scope": "mask-markup",
    "subject": "merging with the develop branch",
    "merge": null,
    "header": "feat(mask-markup): merging with the develop branch",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "0153c1aba8a08ba6c974fcc69476a87b9086f6c5",
    "gitTags": " (develop)",
    "committerDate": "2019-05-02 07:12:59 +0300",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@1.2.0"
  },
  {
    "type": "fix",
    "scope": "math-toolbar",
    "subject": "math styling fixes",
    "merge": null,
    "header": "fix(math-toolbar): math styling fixes",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "c97a9486af79fc887774994b92024db9117a621c",
    "gitTags": "",
    "committerDate": "2019-05-08 09:53:17 +0300",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@1.2.2"
  },
  {
    "type": "fix",
    "scope": "math-toolbar",
    "subject": "small styling fix for z-index",
    "merge": null,
    "header": "fix(math-toolbar): small styling fix for z-index",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "bf46cc42efb98384b95ced4a9869944ec8d8e782",
    "gitTags": " (origin/fix/math-inline-related-fixes)",
    "committerDate": "2019-05-13 15:17:00 +0300",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@1.2.4"
  },
  {
    "type": "chore",
    "scope": null,
    "subject": "autoFocus prop for automatically focusing math toolbar",
    "merge": null,
    "header": "chore: autoFocus prop for automatically focusing math toolbar",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "c600e24019914ffeec4058cafe6824262001e5db",
    "gitTags": "",
    "committerDate": "2019-05-15 16:31:20 +0300",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@1.3.0"
  },
  {
    "type": "feat",
    "scope": "math-toolbar",
    "subject": "input will automatically grab focus",
    "merge": null,
    "header": "feat(math-toolbar): input will automatically grab focus",
    "body": null,
    "footer": null,
    "notes": [],
    "hash": "2a344fb4f61e99321bda03538b57251898471497",
    "gitTags": "",
    "committerDate": "2019-05-15 16:23:52 +0300",
    "isTagged": true,
    "tag": "@pie-lib/math-toolbar@1.3.0"
  }
]