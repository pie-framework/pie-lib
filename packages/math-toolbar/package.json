{
  "name": "@pie-lib/math-toolbar",
  "publishConfig": {
    "access": "public"
  },
  "version": "0.5.1",
  "description": "Math toolbar for editing math equations",
  "keywords": [
    "math",
    "toolbar",
    "equations",
    "expressions"
  ],
  "repository": "pie-framework/pie-lib",
  "main": "lib/index.js",
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@material-ui/core": "^3.8.3",
    "@material-ui/icons": "^3.0.2",
    "@pie-lib/math-input": "^4.6.0",
    "classnames": "^2.2.6",
    "debug": "^4.1.1",
    "keycode": "^2.2.0",
    "mathquill": "git+https://github.com/pie-framework/mathquill-webpack.git",
    "prop-types": "^15.6.2"
  },
  "peerDependencies": {
    "react": "^16.7.0",
    "react-dom": "^16.7.0"
  },
  "devDependencies": {
    "@pie-lib/test-utils": "^0.0.3"
  }
}
