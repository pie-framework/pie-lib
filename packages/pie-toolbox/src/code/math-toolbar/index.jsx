import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import EditorAndPad from './editor-and-pad';
import { DoneButton } from './done-button';
import { withStyles } from '@material-ui/core/styles';
import MathPreview from './math-preview';

export { MathPreview };

export class MathToolbar extends React.Component {
  static propTypes = {
    autoFocus: PropTypes.bool,
    allowAnswerBlock: PropTypes.bool,
    controlledKeypad: PropTypes.bool,
    controlledKeypadMode: PropTypes.bool,
    keypadMode: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    classNames: PropTypes.object,
    error: PropTypes.string,
    maxResponseAreas: PropTypes.number,
    showKeypad: PropTypes.bool,
    noDecimal: PropTypes.bool,
    additionalKeys: PropTypes.array,
    latex: PropTypes.string.isRequired,
    onAnswerBlockAdd: PropTypes.func,
    onChange: PropTypes.func,
    onDone: PropTypes.func.isRequired,
    onFocus: PropTypes.func,
    onBlur: PropTypes.func,
    hideDoneButton: PropTypes.bool,
  };

  static defaultProps = {
    classNames: {},
    keypadMode: 'item-authoring',
    autoFocus: false,
    allowAnswerBlock: false,
    controlledKeypad: false,
    controlledKeypadMode: false,
    noDecimal: false,
    showKeypad: true,
    additionalKeys: [],
    onChange: () => {},
    onAnswerBlockAdd: () => {},
    onFocus: () => {},
    hideDoneButton: false,
  };

  constructor(props) {
    super(props);
    this.state = {
      latex: props.latex,
    };
  }

  done = () => {
    this.props.onDone(this.state.latex);
  };

  UNSAFE_componentWillReceiveProps(nextProps) {
    this.setState({ latex: nextProps.latex });
  }

  onChange = (latex) => {
    this.setState({ latex });
    this.props.onChange(latex);
  };

  render() {
    const { latex } = this.state;
    const {
      classNames,
      autoFocus,
      allowAnswerBlock,
      onAnswerBlockAdd,
      controlledKeypad,
      controlledKeypadMode,
      keypadMode,
      noDecimal,
      additionalKeys,
      showKeypad,
      onFocus,
      onBlur,
      hideDoneButton,
      error,
      maxResponseAreas,
    } = this.props;

    return (
      <PureToolbar
        autoFocus={autoFocus}
        classNames={classNames}
        onAnswerBlockAdd={onAnswerBlockAdd}
        allowAnswerBlock={allowAnswerBlock}
        latex={latex}
        additionalKeys={additionalKeys}
        noDecimal={noDecimal}
        keypadMode={keypadMode}
        onChange={this.onChange}
        onDone={this.done}
        onFocus={onFocus}
        onBlur={onBlur}
        showKeypad={showKeypad}
        controlledKeypad={controlledKeypad}
        controlledKeypadMode={controlledKeypadMode}
        hideDoneButton={hideDoneButton}
        error={error}
        maxResponseAreas={maxResponseAreas}
      />
    );
  }
}

export class RawPureToolbar extends React.Component {
  static propTypes = {
    classNames: PropTypes.object,
    latex: PropTypes.string.isRequired,
    keypadMode: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    hideInput: PropTypes.bool,
    noLatexHandling: PropTypes.bool,
    layoutForKeyPad: PropTypes.object,
    onChange: PropTypes.func.isRequired,
    onDone: PropTypes.func.isRequired,
    onBlur: PropTypes.func,
    onAnswerBlockAdd: PropTypes.func,
    additionalKeys: PropTypes.array,
    onFocus: PropTypes.func,
    classes: PropTypes.object.isRequired,
    autoFocus: PropTypes.bool,
    noDecimal: PropTypes.bool,
    allowAnswerBlock: PropTypes.bool,
    controlledKeypad: PropTypes.bool,
    controlledKeypadMode: PropTypes.bool,
    showKeypad: PropTypes.bool,
    hideDoneButton: PropTypes.bool,
    hideDoneButtonBackground: PropTypes.bool,
    error: PropTypes.any,
    maxResponseAreas: PropTypes.number,
  };

  static defaultProps = {
    classNames: {},
    hideDoneButtonBackground: false,
  };

  render() {
    const {
      classNames,
      autoFocus,
      allowAnswerBlock,
      onAnswerBlockAdd,
      controlledKeypad,
      controlledKeypadMode,
      additionalKeys,
      showKeypad,
      keypadMode,
      noDecimal,
      hideInput,
      noLatexHandling,
      layoutForKeyPad,
      latex,
      onChange,
      onDone,
      onFocus,
      onBlur,
      hideDoneButton,
      hideDoneButtonBackground,
      classes,
      error,
      maxResponseAreas,
    } = this.props;

    return (
      <div className={cx(classes.pureToolbar, (classNames || {}).toolbar)}>
        <div />
        <EditorAndPad
          autoFocus={autoFocus}
          keypadMode={keypadMode}
          classNames={classNames || {}}
          controlledKeypad={controlledKeypad}
          controlledKeypadMode={controlledKeypadMode}
          noDecimal={noDecimal}
          hideInput={hideInput}
          noLatexHandling={noLatexHandling}
          layoutForKeyPad={layoutForKeyPad}
          showKeypad={showKeypad}
          additionalKeys={additionalKeys}
          allowAnswerBlock={allowAnswerBlock}
          onAnswerBlockAdd={onAnswerBlockAdd}
          latex={latex}
          onChange={onChange}
          onFocus={onFocus}
          onBlur={onBlur}
          error={error}
          maxResponseAreas={maxResponseAreas}
        />
        {(!controlledKeypad || (controlledKeypad && showKeypad)) && !hideDoneButton && (
          <DoneButton hideBackground={hideDoneButtonBackground} onClick={onDone} />
        )}
      </div>
    );
  }
}
const styles = () => ({
  pureToolbar: {
    display: 'flex',
    width: '100%',
    zIndex: 8,
    alignItems: 'center',
  },
});

export const PureToolbar = withStyles(styles)(RawPureToolbar);
