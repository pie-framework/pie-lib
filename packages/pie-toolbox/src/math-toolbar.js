import {
  MathPreview,
  MathToolbar,
  RawPureToolbar,
  PureToolbar,
} from './code/math-toolbar';

export {
  MathPreview,
  MathToolbar,
  RawPureToolbar,
  PureToolbar,
};
