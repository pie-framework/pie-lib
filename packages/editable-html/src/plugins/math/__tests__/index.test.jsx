import React from 'react';
import debug from 'debug';
import MockChange from '../../image/__tests__/mock-change';
import { Data } from 'slate';
import MathPlugin, { serialization, inlineMath, CustomToolbarComp } from '../index';
import { shallow } from 'enzyme';
import { MathToolbar } from '@pie-lib/math-toolbar';
jest.mock('@pie-framework/mathquill', () => ({
  StaticMath: jest.fn(),
  getInterface: jest.fn().mockReturnThis(),
  registerEmbed: jest.fn()
}));
jest.mock('@pie-lib/math-toolbar', () => ({
  MathPreview: () => <div />,
  MathToolbar: () => <div />
}));
const log = debug('@pie-lib:editable-html:test:math');

// I believe @andrei is moving this stuff out.
describe('MathPlugin', () => {
  describe('toolbar', () => {
    describe('onClick', () => {
      let plugin, mockChange, value, onChange;
      beforeEach(() => {
        plugin = MathPlugin({});
        mockChange = new MockChange();
        value = {
          change: jest.fn(() => mockChange)
        };
        onChange = jest.fn();
        plugin.toolbar.onClick(value, onChange);
      });

      test('calls insertInline', () => {
        expect(mockChange.insertInline).toBeCalledWith(
          expect.objectContaining({ data: inlineMath().data })
        );
      });

      test('it calls onChange', () => {
        expect(onChange).toHaveBeenCalledWith(mockChange);
      });
    });
  });

  describe('renderNode', () => {
    test('the component has props', () => {
      const plugin = MathPlugin({});
      const { props } = plugin.renderNode({ node: { type: 'math' } });
      expect(props.node).toEqual({ type: 'math' });
    });
  });

  describe('serialization', () => {
    describe('deserialize', () => {
      const assertDeserialize = (html, expected, wrapType) => {
        it(`innerHTML: ${html} is deserialized to: ${expected} with wrapType: ${wrapType}`, () => {
          const el = {
            tagName: 'span',
            childNodes: [],
            getAttribute: jest.fn(() => ''),
            hasAttribute: jest.fn(() => true),
            innerHTML: html
          };
          const next = jest.fn();

          const out = serialization.deserialize(el, next);
          expect(out).toEqual({
            object: 'inline',
            type: 'math',
            isVoid: true,
            nodes: [],
            data: {
              latex: expected,
              wrapper: wrapType
            }
          });
        });
      };

      assertDeserialize('$$&lt;$$', '<', MathPlugin.DOLLAR);
      assertDeserialize('$&lt;$', '<', MathPlugin.DOLLAR);
      assertDeserialize('\\(&lt;\\)', '<', MathPlugin.ROUND_BRACKETS);
      assertDeserialize('\\[&lt;\\]', '<', MathPlugin.ROUND_BRACKETS);
      assertDeserialize('latex', 'latex', MathPlugin.ROUND_BRACKETS);
      assertDeserialize('\\displaystyle foo', 'foo', MathPlugin.ROUND_BRACKETS);
    });

    describe('serialize', () => {
      const assertSerialize = (latex, expectedHtml, wrapper) => {
        wrapper = wrapper || MathPlugin.ROUND_BRACKETS;
        it(`${latex} is serialized to: ${expectedHtml}`, () => {
          const object = {
            kind: 'inline',
            type: 'math',
            isVoid: true,
            nodes: [],
            data: Data.create({ latex, wrapper })
          };
          const children = [];

          const out = serialization.serialize(object, children);
          log('out: ', out);
          expect(out).toEqual(
            <span data-latex="" data-raw={latex}>
              {expectedHtml}
            </span>
          );
        });
      };

      assertSerialize('latex', '\\(latex\\)', MathPlugin.ROUND_BRACKETS);
      assertSerialize('latex', '\\(latex\\)', MathPlugin.SQUARE_BRACKETS);
      assertSerialize('latex', '$latex$', MathPlugin.DOLLAR);
      assertSerialize('latex', '$latex$', MathPlugin.DOUBLE_DOLLAR);

      /**
       * Note that when this is converted to html it get's escaped - but that's an issue with the slate html-serializer.
       */
      assertSerialize('<', '\\(<\\)');
    });
  });
});

describe('CustomToolbarComp', () => {
  let onDataChange;
  let onToolbarDone;

  const wrapper = extras => {
    let mockChange = new MockChange();
    const defaults = {
      node: {
        key: '1',
        data: Data.create({ latex: 'foo' }),
        equals: () => true
      },
      value: {
        document: {
          getNextText: jest.fn().mockReturnValue({ key: 'nt' })
        },
        change: jest.fn().mockReturnValue(mockChange)
      },
      onDataChange,
      onToolbarDone
    };

    const props = {
      ...defaults,
      ...extras
    };

    return shallow(<CustomToolbarComp {...props} />);
  };

  describe('render', () => {
    it('renders without default keypadMode', () => {
      const w = wrapper();

      expect(w).toMatchSnapshot();

      w.setProps({ pluginProps: { math: { keypadMode: 3 } } });

      expect(w).toMatchSnapshot();
    });

    it('renders with default keypadMode', () => {
      const w = wrapper({ pluginProps: { math: { keypadMode: 'geometry' } } });

      expect(w).toMatchSnapshot();

      w.setProps({ pluginProps: { math: { keypadMode: 3 } } });

      expect(w).toMatchSnapshot();
    });
  });

  describe('onDone', () => {
    it('calls onToolbarDone', () => {
      onToolbarDone = jest.fn();
      const w = wrapper();
      w.find(MathToolbar).prop('onDone')('oo');
      expect(onToolbarDone).toHaveBeenCalledWith(expect.anything(), false);
    });
  });

  describe('onChange', () => {
    it('calls onDataChange', () => {
      onDataChange = jest.fn();
      const w = wrapper();
      w.find(MathToolbar).prop('onChange')('oo');
      expect(onDataChange).toHaveBeenCalledWith('1', { latex: 'oo' });
    });
  });
});
