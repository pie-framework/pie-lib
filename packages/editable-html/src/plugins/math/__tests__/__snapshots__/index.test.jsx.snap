// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomToolbarComp render renders with default keypadMode 1`] = `
<MathToolbar
  autoFocus={true}
  keypadMode="geometry"
  latex="foo"
  onChange={[Function]}
  onDone={[Function]}
/>
`;

exports[`CustomToolbarComp render renders with default keypadMode 2`] = `
<MathToolbar
  autoFocus={true}
  keypadMode={3}
  latex="foo"
  onChange={[Function]}
  onDone={[Function]}
/>
`;

exports[`CustomToolbarComp render renders without default keypadMode 1`] = `
<MathToolbar
  autoFocus={true}
  latex="foo"
  onChange={[Function]}
  onDone={[Function]}
/>
`;

exports[`CustomToolbarComp render renders without default keypadMode 2`] = `
<MathToolbar
  autoFocus={true}
  keypadMode={3}
  latex="foo"
  onChange={[Function]}
  onDone={[Function]}
/>
`;
